samples/edgehog_app/src/main.c

```diff
@@ -6,358 +6,458 @@
+        k_sleep(K_MSEC(100));
     }
+
+    eres = edgehog_device_start(edgehog_device);
+    if (eres != EDGEHOG_RESULT_OK) {
+        LOG_ERR("Unable to start edgehog device"); // NOLINT
+        return;
+    }
+
+    while (!atomic_test_bit(&device_threads_flags, DEVICE_THREADS_FLAGS_TERMINATION)) {
+        k_sleep(K_MSEC(EDGEHOG_DEVICE_PERIOD_MS));
+    }
+
+    LOG_INF("End of sample, Edgehog stopping imminent."); // NOLINT
+    eres = edgehog_device_stop(edgehog_device, K_FOREVER);
+    if (eres != EDGEHOG_RESULT_OK) {
+        LOG_ERR("Unable to stop the edgehog device"); // NOLINT
+        return;
+    }
+
+    LOG_INF("Edgehog device will now be destroyed."); // NOLINT
+    edgehog_device_destroy(edgehog_device);
+
+    LOG_INF("Edgehog thread will now be terminated."); // NOLINT
+    k_sleep(K_MSEC(MSEC_PER_SEC));
 }
 
-static void datastream_object_events_handler(astarte_device_datastream_object_event_t event)
+static void astarte_device_thread_entry_point(void *arg1, void *arg2, void *arg3)
 {
-    edgehog_device_handle_t *edgehog_device
-        = (edgehog_device_handle_t *) event.data_event.user_data;
+    ARG_UNUSED(arg1);
+    ARG_UNUSED(arg2);
+    ARG_UNUSED(arg3);
 
-    edgehog_device_datastream_object_events_handler(*edgehog_device, event);
+    astarte_result_t ares = ASTARTE_RESULT_OK;
+
+    // Initialize the Astarte device
+    char cred_secr[ASTARTE_PAIRING_CRED_SECR_LEN + 1] = CONFIG_ASTARTE_CREDENTIAL_SECRET;
+    char device_id[ASTARTE_DEVICE_ID_LEN + 1] = CONFIG_ASTARTE_DEVICE_ID;
+
+    astarte_device_config_t astarte_device_config = { 0 };
+    memset(&astarte_device_config, 0, sizeof(astarte_device_config));
+    astarte_device_config.http_timeout_ms = HTTP_TIMEOUT_MS;
+    astarte_device_config.mqtt_connection_timeout_ms = MQTT_FIRST_POLL_TIMEOUT_MS;
+    astarte_device_config.mqtt_poll_timeout_ms = MQTT_POLL_TIMEOUT_MS;
+    astarte_device_config.connection_cbk = astarte_device_connection_callback;
+    astarte_device_config.disconnection_cbk = astarte_device_disconnection_callback;
+    astarte_device_config.datastream_individual_cbk = astarte_device_datastream_individual_callback;
+    astarte_device_config.datastream_object_cbk = astarte_device_datastream_object_callback;
+    astarte_device_config.property_set_cbk = astarte_device_property_set_callback;
+    astarte_device_config.property_unset_cbk = astarte_device_property_unset_callback;
+    astarte_device_config.cbk_user_data = (void *) &edgehog_device;
+    memcpy(astarte_device_config.cred_secr, cred_secr, sizeof(cred_secr));
+    memcpy(astarte_device_config.device_id, device_id, sizeof(device_id));
+
+    ares = astarte_device_new(&astarte_device_config, &astarte_device);
+    if (ares != ASTARTE_RESULT_OK) {
+        LOG_ERR("Astarte device creation failure."); // NOLINT
+        return;
+    }
+
+    // Signal the Edgehog thread to create the device
+    atomic_set_bit(&device_threads_flags, DEVICE_THREADS_FLAGS_CREATE_EDGEHOG);
```

Comment:
Is there a reason why you changed your mind from [this comment](https://github.com/astarte-platform/astarte-device-sdk-zephyr/pull/192#discussion_r1795366046)?

Reply:
It's more that when thread A starts thread B, it should be thread A that stops it (B) IMO, that is to avoid having a big mess in which you have no idea where a thread is started or stopped. Instead when keeping the two operations on the same thread is just a matter of finding them. For our small example, it does not matter much but it's probably a good practice for even for simple applications.